WalletWasabi/Crypto/Transcript.cs

```diff
@@ -58,32 +66,32 @@ public IEnumerable<Scalar> GenerateSecretNonces(IEnumerable<Scalar> secrets, Was
 				forked.Key(secret.ToBytes(), false);
 			}
 
-			// add additional randomness
+			// Add additional randomness
 			forked.Key(random.GetBytes(KeySizeInBytes), false);
 
-			// FIXME for the general case we need publicPoints.Count() * Witness.Length
-			// secret nonces per statement.
-			// this method should return a delegate here so that the following lines
-			// can be used repeatedly, or given the number of public inputs it could
-			// just return IEnumerable<IEnumerable<Scalar>> which is probably uglier.
-
-			// generate a new scalar for each secret using this updated state as a seed
-			return Enumerable
-				.Range(0, secrets.Count())
-				.Select( _ => forked.Prf(KeySizeInBytes, false))
-				.Select( rnd => new Scalar(rnd) );
+			IEnumerable<Scalar> NoncesGenerator()
+			{
+				while (true)
+				{
+					yield return new Scalar(forked.Prf(KeySizeInBytes, false));
+				}
+			};
+
+			// Generate a new scalar for each secret using this updated state as a seed.
+			return NoncesGenerator;
 		}
-
+		
 		public void CommitPublicNonces(IEnumerable<GroupElement> nonces)
 		{
-			// FIXME loop Guard.False($"{nameof(nonce)}.{nameof(nonce.IsInfinity)}", nonce.IsInfinity);
+			CryptoGuard.NotInfinity(nameof(nonces), nonces);
 			AddMessages(NonceTag, nonces.Select(x => x.ToBytes()));
 		}
 
-		// generate Fiat Shamir challenges
+		// Generate Fiat Shamir challenges
 		public Scalar GenerateChallenge()
 		{
 			_strobe.AddAssociatedMetaData(ChallengeTag, false);
+			_strobe.AddAssociatedMetaData(BitConverter.GetBytes(KeySizeInBytes), true); // TODO: does this make sense?
```

Comment:
doesn't hurt but since it's constant and can be hard coded it's not really necessary either and could be argued to be implied by the challenge tag.

note that it's not portable: from my understanding `BitConverter` only provides conversion in native byte order, and I couldn't find a standard library function that guarantees it but was too lazy to check if little endian is true and reverse the bytes when I sketched this stuff out